bitcoin/src/bip32.rs

```diff
@@ -133,51 +127,60 @@ pub struct NormalChildNumber(pub u32);
 
 /// A iterator over the children of a DerivationPath trying to convert them into NormalChildNumber
 /// made only to be used internally in `derive_pub()`
-pub struct NormalChildIterator {
+pub struct NormalDerivationPath {
     base: Vec<NormalChildNumber>,
 }
-impl Into<NormalChildIterator> for DerivationPath {
-    fn into(self) -> NormalChildIterator {
-        NormalChildIterator {
-            base: self
-                .0
-                .into_iter()
-                .filter_map(|child| {
-                    if let ChildNumber::Normal { index } = child {
-                        Some(NormalChildNumber(index))
-                    } else {
-                        None
-                    }
-                })
-                .collect(),
+
+impl DerivationPath {
+    /// Try to convert this DerivationPath into a NormalDerivationPath
+    pub fn try_into_normal(&self) -> Result<NormalDerivationPath, Error> {
+        NormalDerivationPath::try_from(self)
+    }
+}
+
+impl TryFrom<&DerivationPath> for NormalDerivationPath {
+    type Error = Error;
+    fn try_from(value: &DerivationPath) -> Result<Self, Self::Error> {
+        let mut base = Vec::with_capacity(value.0.len());
+        for child in value.0.iter() {
+            match child {
+                ChildNumber::Normal(index) => {
+                    base.push(NormalChildNumber(*index));
+                }
+                ChildNumber::Hardened(_) => {
+                    return Err(Error::CannotDeriveFromHardenedKey);
+                }
+            }
         }
+        Ok(NormalDerivationPath { base })
     }
 }
-impl Iterator for NormalChildIterator {
+
+impl Iterator for NormalDerivationPath {
     type Item = NormalChildNumber;
     fn next(&mut self) -> Option<Self::Item> { self.base.pop() }
```

Comment:
No, this is still wrong. It needs to implement `IntoIterator`, not `Iterator`.

Reply:
Why ?

Reply:
`Iterator` does not consume `self` and yields references `&T`. Whereas `IntoIterator` may consume and may yield any of `T`, `&T` or `&mut T`, depending on the context.
Your `.pop()` returns `T`.

Reply:
Oh ok. Done

Reply:
`pop` also returns them in reverse which is completely wrong.

Reply:
Actually not because we initiate the vec with push

Reply:
That's super confusing. It's best to avoid "being clever" like this.

Reply:
@Kixunil without looking at this specific case, it's fairly common to implement iterators by using `push` to construct and `pop` to yield. Yes, it is potentially confusing, but it's way more efficient than any alternate way of using `Vec`, and it's reasonably common (I think this appears in rust-miniscript as well).

You could alternately use `VecDeque` if you want to be explicit.

Reply:
I found `VecDeque` while trying to solve this, but look a little bloated for the case.
We just need to guarantee that the order of the elements is right as the expected.

> Yes, it is potentially confusing, but it's way more efficient than any alternate way of using Vec.

Actually not confusing enough for someone to not understand, on my opinion... we can consider that the one whos reading is a developer.

Reply:
@apoelstra you can just wrap `alloc::vec::IntoIter` if the inner type is already an iterator or use `core::slice::Iter` for borrowed iterators. I've never seen people hacking it using `push` and `pop`.

Reply:
@Kixunil [IntoIter example](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) from the guys who created the language. My implementation ends being coincidentally like the official one.

Reply:
Lol, OK, an academic example. Anyway, it was already decided that we want idiomatic code, so let's do that.

> we can consider that the one whos reading is a developer.

Specifically, a maintainer - someone who takes care of the code regularly. And as one I wouldn't want to maintain confusing code because it takes longer and leads to more bugs. Why is this not obvious?

Reply:
We cannot just wraps it as if we are just changing owners of the vector because we need to check every single `ChildNumber`. Lacks a example where we can do it, if you find some i would gladly implement it for this PR.

>Specifically, a maintainer - someone who takes care of the code **regularly.**

Kinda, there is other developers that use the code and needs to read it.

>And as one I wouldn't want to maintain confusing code because it takes longer and leads to more bugs.

Is `push()` confusing ? Skill Issue. Is literally how they expect us to work in this context.

Don't get me wrong, my point is trying to understand a better way to doing it and why my way is worst. Just because we can do it using "lower level tools" doesnt mean that will be better and mainly "readable" or  "idiomatic" as you would argue. 

With this disagreement precisely with syntax of the method. I would like the opinion of others developers too and to ensure that the problem is with the code.
@storopoli 
@apoelstra

Reply:
> we need to check every single `ChildNumber`

We do that upfront to create `NormalDerivationPath`, so there's nothing to check in the `IntoIterator::IntoIter`.

> Lacks a example where we can do it

Check how `String::from_utf8` and `core::str::Chars` work.

> Kinda, there is other developers that use the code and needs to read it.

Those don't need to read it, they just need to see the docs.

> Skill Issue.

If this is your approach to programming you need to seriously rethink it or stop doing so. Especially security critical stuff like Bitcoin. We need the code to be as obvious as possible if we're not gaining anything. You `push`/`pop` hack is not getting us any interesting property that cannot be achieved through `vec::IntoIter` or other means. Also I think you should respect more that you're a newbie in Rust and secure software.

>  why my way is worst

I believe I've already explained it.

* Forces unneeded copies (but sometimes they are unavoidable so having a type is still useful)
* Stores stuff in reverse order to the natural one without justification
* Prevents future improvements such as implementing `AsRef<[NormalChildNumber]>`

> "lower level tools"

Iterators absolutely aren't lower level. They are standard high-level Rust constructs. If you don't know them properly you really need to study the language more.

Reply:
@jaoleal everyone maintaining this codebase has limited time, and we place a premium on readability. In this case I doubt there is any measurable performance benefit to using push/pop versus just wrapping a `vec::IntoIter`.

As an aside, @Kixunil is the only person taking an active interest in this PR, which judging from the number of comments (often about basic correctness issues) seems like it needs a lot of guidance. Comments like "Skill issue." read as snarky and unappreciative. I'm sure that Kix could rewrite this whole PR in less time than this review has taken. Skill is not the issue here, but how much of a maintainability cost any new code brings with it.

Reply:
>As an aside, @Kixunil is the only person taking an active interest in this PR, which judging from the number of comments (often about basic correctness issues) seems like it needs a lot of guidance.

Indeed. That one of the reasons that i tagged him at the beginning of this draft. Please accept my honest apologies for making a joke, but was forced after received laugh as a response. 
I appreciate the high standard code that `rust-bitcoin` has, being the main reason why i'm really focusing on this PR  lately... learning a lot. That's why i deeply want to understand reasons behind code decisions that @Kixunil is requesting rather to trying to guess the meaning.

>In this case I doubt there is any measurable performance benefit to using push/pop versus just wrapping a vec::IntoIter

>Iterators absolutely aren't lower level. 

Another mistake of mine i guess. I was not talking about using `IntoIter`, actually about the initial conversion beetween `DerivationPath` and  `NormalDerivationPath` using allocators. [Context.](https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1681520701).
I think that i was clear while agreeing using `IntoIter` after commenting ["Ok Done"](https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1679545889), I blame it on my lack of my experience with another language than portuguese. This is actually already commited on my code after the explanation from [storopoli](https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1679498571).

I didnt pushed anything yet because of [this](https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1678955630) requested change that is taking more time than i think to reach the `rust-bitcoin` standard.

>everyone maintaining this codebase has limited time,

Of course, never had intention of wasting time of anybody since I reserving mine to help. I just ask more precise requests comments like these:
https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1679498571
https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1678957618
https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1678951635
https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1677390376
Most of them by Kix. Without any Irony or presupposition of knowledge of the internal ideal of code.

This makes we all spend more time and push away developers like me that got things on the run and are trying to contribute. I do not see why not adopting new contributors since even a code from a newbie are improving the codebase. But that's not my decision to take.

Reply:
Cultural differences I guess? The "LOL" was actually me laughing at myself for being refuted on technicality (if you've seen enough of my comments you'll see I like being technically correct but I also accept fair critique). Anyway, let's put these things aside.

The real question is how much time do we have to mentor Rust newbies? I don't have much and I'd really love other things to move forward but I'm open to giving people a chance so I review anyway.

Here are some topics you probably need to learn:

* newtypes and how they protect invariants (see also our lock time types for practical examples)
* iterators, `IntoIterator` trait and the standard APIs like `iter()`
* error handling
* heap allocations - what are they, what the cost is and how to avoid them if possible

Also feel free to browse our issues mentioning BIP32/child numbers. We've already discussed bunch of things regarding newtypes and such.